site/learn/expert-advice/authentication/webauthn-explained.md

```diff
@@ -0,0 +1,478 @@
+
+You again have the `publicKey` as the top level options field, indicating that you are using WebAuthn.
+
+Then there is the `rp` object, which has information about the website that is requesting the registration operation. This includes the name to be displayed to the end user by the authenticator.
+
+You also have the `user` field, which is information used to tie the user to the credential. Additionally, to create the key, you need to specify the algorithm to use. Do this by passing the correct parameters in `pubKeyCredParams`. Above, the value is `-7`. This is an item in the [IANA COSE Algorithms registry](https://www.iana.org/assignments/cose/cose.xhtml).
+
+After creating the options object, call the `create` method on the credentials API and if it succeeds, the user will be registered with cosmosclownstore.com.
+
+```javascript
+navigator.credentials.create(options)
+  .then(function (newCredentialInfo) {
+    // Send credential info to server 
+  }).catch(function (err) {
+    // No acceptable authenticator or user refused consent. 
+  });
+```
+
+Here's an example response.
+
+```javascript
+{
+  response: {
+    clientDataJSON: { 
+      type: “webauthn.create”,
+      challenge: ... // should match initial challenge
+      origin: ... //
+    },
+    attestationObject: {
+      authData: {
+        attestedCredentialData: {
+          credentialId: ..., 
+          credentialPublicKey: ...,
+        }
+      }, // other stuff about attestation 
+    }
+  }
+}
+```
+
+The `attestationObject` contains items such as:
+
+- the credential Id 
+- the public key 
+
+Store the `credentialId` to identify the user and the public key used when verifying the assertions generated by the other ceremonies. You must also verify the `challenge` and `type` in the `clientDataJSON` object.
+
+You may have noticed that at registration, the website passes down user information. This implies that you know who the user is. Typically, you can require another form of authentication first (i.e. user and password) and then the user will associate a WebAuth credential with this account to make their future authentication easier.
+
+As the relying party (website), you may specify what type of authenticators you want to support. You can’t get too granular, but you can specify if you only want to support platform authenticators or cross-platform authenticators.
+
+Next, let's look at the attestation ceremony.
+
+## The Attestation Ceremony
+
+Attestation is the third major ceremony and the question it answers is: How do I know this authenticator the user is using is legitimate? It does this through a chain of signatures.
+
+Each type of authenticator has a shared private key. This key is totally separate from private keys used for the authentication and registration ceremonies, which are tied to individual websites. The corresponding public key is stored in a public metadata registry. Here's the [FIDO alliance metadata service](https://fidoalliance.org/metadata/).
```

Comment:
I'm having trouble finding the source now, but I believe that authenticator manufacturers must pay a fee to have their device listed as part of the metadata service. Not all devices are listed (e.g. I don't think any Apple devices are in the blob).

I am not sure how certificate chain validation works when an authenticator is not listed in the metadata service. I suppose it resolves to some trusted root CA?

Reply:
Looks like Apple provides a root public cert: https://groups.google.com/a/fidoalliance.org/g/fido-dev/c/3J3lIGwRLmA